### PR TITLE
Allow more efficient search by using WPorg API to fech results.

### PR DESCRIPTION
**Pull request by louim**
This PR bring more useful search to help with issues like #76. It is based on @Rarst WP org API wrapper. I had to change the search to look only for plugins or themes, not the two combined.

The query is made to the WP org API, results are then matched to entries of the database. WP org results are used as the reference for the sort order to prevent any oddities in the pagination.

I added a link on the plugin / theme name to the corresponding Wordpress page. Makes it easier to check plugin descriptions / information. 

This PR also open the way for more search option in the future if we want to (like tags or popular from the WP site).

I updated the dependencies (was needed for the DB lookup).

_NOTE_ I'm making use of `array_column` which is PHP 5.5+ . wpackagist.org is currently running 5.4.35 which EOL soon. Time to upgrade @tamlyn? :wink:

### CODE REVIEW

**Comment by louim** [comment]
Ping! @tamlyn @lavoiesl

File: web/index.php [hunk 1]
```diff
@@ -108,43 +105,47 @@
...
-            ->addOrderBy('name', 'ASC')
-            ->setParameter(':name', "%{$query}%")
-            ->setParameter(':order', "{$query}%");
+            ->andWhere('p.name IN (:name)')
+            ->setParameter(':name', $slugs, \Doctrine\DBAL\Connection::PARAM_STR_ARRAY);
+
+        // This gymnastic is needed because the SQLITE database doesn't support ordering by FIELD, so we need to
+        // do it in PHP. We order the results from the database to match what the  wordpress API is returning us.
+        $statement = $queryBuilder->execute();
+        $packageResults = $statement->fetchAll(PDO::FETCH_ASSOC);
```
Comment (tamlyn) [inline web/index.php]: This always comes back empty for me. If I manually build up the query it works. From stepping into the code it looks like it's trying to bind the array value to the PDO statement without unwrapping it first. Any idea what's going on?

Comment (tamlyn) [inline web/index.php]: I also get the following `Notice: Array to string conversion in /vagrant/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php on line 628`

Comment (louim) [inline web/index.php]: Are you using PHP 5.5?

File: composer.json [hunk 2]
```diff
@@ -15,7 +14,7 @@
         "symfony/translation": "~2.3",
         "pagerfanta/pagerfanta": "dev-master",
         "franmomu/silex-pagerfanta-provider": "dev-master",
-        "doctrine/dbal": "2.2.*",
+        "doctrine/dbal": "v2.5.1",
```
Comment (lavoiesl) [inline composer.json]: Why are you bumping the dbal version, why are you requiring `2.5.1` specifically instead of `~2.5`, and why did you add a `v` while all the others don't use it?

File: web/index.php [hunk 3]
```diff
@@ -58,17 +59,10 @@
     ->add('q', 'search')
     ->add('type', 'choice', array(
         'choices' => array(
-            'any'     => 'All packages',
             'plugin'  => 'Plugins',
             'theme'   => 'Themes',
         ),
     ))
-//    ->add('active_only', 'choice', array(
```
Comment (lavoiesl) [inline web/index.php]: Why is this commented instead of removed?

File: web/index.php [hunk 4]
```diff
@@ -88,9 +82,12 @@
 $app->get('/search', function (Request $request) use ($app, $searchForm) {
     /** @var \Doctrine\DBAL\Query\QueryBuilder $queryBuilder */
     $queryBuilder = $app['db']->createQueryBuilder();
+    $wporgClient  = \Rarst\Guzzle\WporgClient::getClient();
```
Comment (lavoiesl) [inline web/index.php]: Why is this FQDN instead of using `use` at the top?

File: web/index.php [hunk 5]
```diff
@@ -88,9 +82,12 @@
...
+    $currentPage  = $request->query->get('page', 1);
     $type         = $request->get('type');
     $active       = $request->get('active_only');
```
Comment (lavoiesl) [inline web/index.php]: If you are removing the `active_only` option, you should remove this line as well.

File: web/index.php [hunk 6]
```diff
@@ -108,43 +105,47 @@
...
-    }
-
     if (!empty($query)) {
+        // We match the results from the wordpress API to our database.
         $queryBuilder
-            ->andWhere('name LIKE :name')
-            ->addOrderBy('name LIKE :order', 'DESC')
-            ->addOrderBy('name', 'ASC')
-            ->setParameter(':name', "%{$query}%")
-            ->setParameter(':order', "{$query}%");
```
Comment (lavoiesl) [inline web/index.php]: Not sure about removing the ordering, this means partial matches can now score higher than exact matches

File: web/index.php [hunk 7]
```diff
@@ -108,43 +105,47 @@
...
+        // do it in PHP. We order the results from the database to match what the  wordpress API is returning us.
+        $statement = $queryBuilder->execute();
+        $packageResults = $statement->fetchAll(PDO::FETCH_ASSOC);
+        $packageResults = array_column($packageResults, null, 'name');
+
+        foreach ($slugs as $name) {
+            $results[] = $packageResults[$name];
+        }
+
+        $adapter    = new FixedAdapter($wporgQueryResults['info']['results'], $results);
```
Comment (lavoiesl) [inline web/index.php]: Avoid aligning `=` signs

File: web/index.php [hunk 8]
```diff
@@ -108,43 +105,47 @@
...
     } else {
+        // If we have no search query, we just paginate based on last committed
         $queryBuilder
             ->addOrderBy('last_committed', 'DESC');
+        $adapter    = new DoctrineDbalSingleTableAdapter($queryBuilder, 'p.name');
```
Comment (lavoiesl) [inline web/index.php]: Avoid aligning `=` signs

File: web/templates/search.twig [hunk 9]
```diff
@@ -28,7 +28,7 @@
                         {{ package['class_name'] | format_category }}
                     </td>
                     <td data-name="{{ package['name'] | e }}">
-                        {{ package['name'] | e }}
+                        <a target="_blank" href="https://wordpress.org/{{ package['class_name'] | format_category | lower }}s/{{ package['name'] | e }}">{{ package['name'] | e }}</a>
```
Comment (lavoiesl) [inline web/templates/search.twig]: Use twig syntax `package.class_name`, not `package['class_name']`

File: src/Command/UpdateCommand.php [hunk 10]
```diff
@@ -94,7 +91,11 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
 
             //get versions as [version => url]
-            $versions = $info['versions'] ?: [];
```
Comment (lavoiesl) [inline src/Command/UpdateCommand.php]: Why are you changing this? It doesn't seem to belong in this PR, please open a separate one if there is a need for it.

**Comment by lavoiesl** [comment]
:+1: I like the spirit of this PR, would approve after issues are addressed.

File: composer.json (hunk 2 above)
Comment (louim) [inline composer.json]: Sorry about that, I just used the way it's shown in [Packagist](https://packagist.org/packages/doctrine/dbal#v2.5.1). Seems like they added the `v` from v2.4.0.

Bumping DBAL was required for the 
```
->setParameter(':name', $slugs, \Doctrine\DBAL\Connection::PARAM_STR_ARRAY);
```
 line to work correctly. 

I think the right way to require it would actually be: `"doctrine/dbal": "^v2.5.1"` to stick to [non breaking update](https://getcomposer.org/doc/articles/versions.md#caret).

File: web/index.php (hunk 4 above)
Comment (louim) [inline web/index.php]: Hmm, I think it may be juste copy paste error. Ill check to fix this.

File: web/index.php (hunk 6 above)
Comment (louim) [inline web/index.php]: We don't really have the choice, Results are now based on what's returned from Wordpress API. If we reorder, our results won't match.

File: web/index.php (hunk 7 above)
Comment (louim) [inline web/index.php]: You are right, I think I'll make a pass with PHP-CS-Fixer. Are we following psr2?

File: src/Command/UpdateCommand.php (hunk 10 above)
Comment (louim) [inline src/Command/UpdateCommand.php]: Yep, Should probably go into another PR. I had to rebuild the database completely, and I was getting a notice for each and every package, so I had to fix it.

File: composer.json (hunk 2 above)
Comment (lavoiesl) [inline composer.json]: The `v` is ignored, it is up to the package to do what it wants. However, I would keep the same style throughout the file.

Do you actually need `>=2.5.1`? It seems `~2.5` would do the trick.

File: src/Command/UpdateCommand.php (hunk 10 above)
Comment (lavoiesl) [inline src/Command/UpdateCommand.php]: Open a new one for that as "cleanup" and it will be merged swiftly. Please move the rollingCurl change to it as well.

File: web/index.php (hunk 6 above)
Comment (lavoiesl) [inline web/index.php]: Yes, WP API will return properly sorted, but you actually ignore that sorting when you do `WHERE IN()`. If you can't keep WP's ordering, I would add some basic ordering, even if it's not perfect.

File: web/index.php (hunk 7 above)
Comment (lavoiesl) [inline web/index.php]: I don't remember, but enforcing PSR-2 in your commits without modifying unnecessary lines is a very good start.

File: web/index.php [hunk 11]
```diff
@@ -5,6 +5,7 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 use Pagerfanta\Pagerfanta;
+use Pagerfanta\Adapter\FixedAdapter;
 use Pagerfanta\Adapter\DoctrineDbalSingleTableAdapter;
```
Comment (lavoiesl) [inline web/index.php]: Is this import still used?

File: web/index.php (hunk 1 above)
Comment (tamlyn) [inline web/index.php]: Yes, 5.5.9. The constant and array get passed through to PDO rather than being expanded by the DBAL.

@lavoiesl does it work for you?

File: web/index.php (hunk 3 above)
Comment (tamlyn) [inline web/index.php]: I think I commented it in a previous commit and @louim removed it. Which is fine with me.

File: web/index.php (hunk 6 above)
Comment (louim) [inline web/index.php]: @lavoiesl it's actually the opposite, we get the slugs of the packages from WP API. We get them in the order they are shown on the actual website. We query those slugs from our database, and order them with PHP to match what we got from Wordpress (see the lines just below.)

File: web/templates/search.twig (hunk 9 above)
Comment (louim) [inline web/templates/search.twig]: I just followed what was already everywhere. Should I change the other lines too?

File: web/index.php (hunk 11 above)
Comment (louim) [inline web/index.php]: Yeah, in case there is no search query. See https://github.com/outlandishideas/wpackagist/pull/100/files#diff-edf673da361757883b6f7562c08cf9e0R143

Comment (lavoiesl) [inline web/index.php]: Alright, my bad.

File: web/index.php (hunk 6 above)
Comment (lavoiesl) [inline web/index.php]: Right you are, I missed that comment.

**Comment by NoelLH** [comment]
I think unfortunately too much has changed for this to be feasible to get mergeable without significant re-work. Given how long it's been open I'll close for now.

**Comment by louim** [comment]
All good, it’s been a while since I’ve used WordPress now. I thought about closing it myself but let it open mostly if someone wanted to pick it back up
